Scs strepla

jkretz opened the pull request:
This is my implementation of scoring Strepla for PySoar

GliderGeek commented:
Thanks @jkretz for opening this pull request an making a start, will soon take a look a provide feedback

GliderGeek commented on PySoar/analysis_run.py:
```diff
@@ -0,0 +1,20 @@
+from competitionDay import CompetitionDay
```
What does this file do?

GliderGeek replied:
i mean the analysis_run.py

jkretz replied:
Oh sorry, I tried something and it accidentally made it into a commit. Line can be deleted

GliderGeek commented on PySoar/exportClass.py:
```diff
@@ -51,7 +51,7 @@ def initiate_best_worst(self, settings, no_legs):
         self.file_name = settings.file_name
 
-        self.wb = xlwt.Workbook()  # initialize excel sheet
+        self.wb = xlwt.Workbook(encoding='latin-1')  # initialize excel sheet
```
What is the reason for this encoding and why specifically 'latin-1'?

jkretz replied:
I added this because PySoar crashes during the generation of an excel sheet in case there is an umlaut (ä,ö,ü) in the competition id. Try the following day that did not work for me:
http://www.soaringspot.com/en_gb/qualifikationsmeisterschaft-landau-landau-ebenberg-2016/results/15m-klasse/task-5-on-2016-08-27/daily

jkretz replied:
And latin was the first to got it working with but maybe it is better to use something else

GliderGeek commented on PySoar/generalFunctions.py:
```diff
@@ -213,9 +213,9 @@ def task_url_from_daily(daily_url):
 def url_format_correct(url_string):
-    if url_string[0:26] != "http://www.soaringspot.com":
+    if (url_string[0:26] != "http://www.soaringspot.com" and url_string[0:21] != "http://www.strepla.de") :        
         return 'URL should start with http://www.soaringspot.com'
```
The string should become:
```python
         return 'URL should start with http://www.soaringspot.com or "http://www.strepla.de'
```

jkretz replied:
Yes, and it might also be a good idea to change the description for the entry widget in main_pysoar.py to something like: "Give Soaringspot/scoringStrepla URL"

GliderGeek commented on PySoar/importClass.py:
```diff
@@ -13,24 +14,30 @@ class SoaringSpotImport(object):
     def __init__(self, url, download_progress):
```
The current structure with a SoaringSpotImport is not very suitable for extension to other sites. It might be a good idea to generalize something like a 'ImportClass' which in turn calls specif SoaringSpot / StrePla functions.

jkretz replied:
In fact I also thought about writing a general input interface but I tried to stay as close to the master branch to facilitate the merge

GliderGeek commented:
I have added some first remarks, however i did not have the time to look at the general workflow. My feeling is that my current codebase was very SoaringSpot oriented and that it was difficult to put the Strepla parts in the right place. Now is a good time to generalize with which i'll gladly help out.

jkretz commented:
All in all I tried to keep it as close to your master branch as possible for easier merging. I also think that it is a good idea to generalize the whole ImportClass a little. The question is if maybe one should generally try to implement the task recognition for Strepla into existing analysis or if it is better to stick with the conversion of the task information from the Strepla into the SeeYou format

GliderGeek commented:
I made a start with generalization to enable a cleaner strepla implementation (and possible other formats).

i created a special Taskpoint.from_scs() constructor which can be used for Strepla.

next up are the classes `Task`, `RaceTask` and `AAT` which contain hidden seeyou logic

GliderGeek commented:
@jkretz: in the last commits i have greatly altered the code, generalizing it to enable a better Strepla implementation.

this breaks your current implementation, but does enable a much cleaner extension.

there are a few files in which Strepla logic should be placed. the following list in not complete, but gives a good first start:

- `scs.py/StreplaDaily`: for downloading the correct files from the site
- `taskpoint.py/from_scs()`: for constructing a taskpoint from a strepla task line in the IGC file
- `task.py/taskpoints_from_scs()`: for looping over the strepla task lines and creating taskpoints
- `competition_day.py/read_igc`: read in necessary lines from igc

let me know if this works for you or when you need help

thanks anyway for this incentive to generalize the code, it was in quite bad shape in some places 🙈

jkretz commented:
@GliderGeek I am currently at JWGC in Lithuania as the meteorologist for the German team and I am lacking the time to have a closer look at what you did. Maybe (but hopefully not) there is some bad weather and I can have a look at it. Nevertheless you did great work implementing all the Strepla stuff into PySoar's main workflow instead of using my workaround

GliderGeek commented:
@jkretz: are you planning to continue with this strepla implementation?

jkretz commented:
I might have some time tomorrow, so I will have a look at it

jkretz commented:
@GliderGeek I started with the Strepla implementation and it seems to work for racing tasks and I will start with AAT in the next days. I also put the Strepla task recognition into the PySoar workflow instead of using my workaround. Hopefully I will be finished with everything by the end of this week and I will push the changes when I'm done

GliderGeek commented:
Great to hear, looking forward to your contributions!

On Wednesday, 20 September 2017, jkretz <notifications@github.com> wrote:

> @GliderGeek <https://github.com/glidergeek> I started with the Strepla
> implementation and it seems to work for racing tasks and I will start with
> AAT in the next days. I also put the Strepla task recognition into the
> PySoar workflow instead of using my workaround. Hopefully I will be
> finished with everything by the end of this week and I will push the
> changes when I'm done
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GliderGeek/PySoar/pull/116#issuecomment-330936864>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APXOiaIQfXXaJPuWtj4CtcxlnBtVLDC6ks5skVbAgaJpZM4OboJR>
> .
>

jkretz commented:
@GliderGeek So here is the first running version of the Strepla implementation. I also added a little bugfix to allow for links starting with https. The way I did it might not be the best style as I'm not super experienced with Python, so if you see any stupid or unnecessary stuff please correct it ;)
Maybe it is a good idea to refactor the `read_igc` function in `competitionDay.py` to improve readability and separate the logic for reading the task information out of the igc files. 
Tomorrow I will do a little more testing to see if there are any issues I haven't found yet

GliderGeek commented:
Awesome! Will try to have a look this weekend. Indeed a good idea to remove the reading logic from competitionday.

GliderGeek commented:
@jkretz: thanks for the implementation, looks good overall. I have done some clean up of the code to make it a bit more readible. Are you using an IDE? I can recommend [PyCharm](https://www.jetbrains.com/pycharm/), it gives a lot of suggestions to improve your code and warn for possible errors

Two other things that i noticed:
- when analysing [this](http://www.strepla.de/scs/Public/scoreDay.aspx?cId=448&idDay=6382) competition, the total task distance is off
- the function `Task.taskpoints_from_scs()` takes two inputs: lscs_lines and lscs_lines_tp. but lscs_lines_tp is contained within lscs_lines. It would be better to not have overlap in the inputs.

Lastly i like your suggestion of removing the igc logic from competitionday. i am going to postpone that however, untill i have the time to do a complete refactor, using the [aerofiles](http://aerofiles.readthedocs.io/en/latest/) library for reading the IGC files.

jkretz commented:
@GliderGeek Thanks for the clean up of the code. I also recognized that the distance is off and I will check where this problem stems from. For the `Task.taskpoints_from_scs()` you are right, the `lscs_lines_tp` can be removed if one extracts coordinates of the turnpoints directly in `task.py` which would make sense. I also have been thinking about that but left it that way to better match the overall structure of `Task.taskpoints_from_cuc()`.
In fact I'm using good old emacs as I am not working with python frequently but using the IDE you proposed might be a good idea to be more productive, thanks

jkretz commented:
@GliderGeek I moved the taskpoint extraction and added shortened legs for finish cylinder. The task distance is now only off by less then 1km which should by ok and is probably due to different distance calculations between Strepla and PySoar

GliderGeek commented:
Nice work!
- Do you know if the same distance correction is present for the startpoint and turnpoints?
- Also i saw that there is only a start-line implemented (no start circle). Is this not possible in Strepla?

jkretz commented:
I checked the distance for a task with 3 Tps and a Daec keyhole and the distance differs by approx. 0.7km ([link](http://www.strepla.de/scs/Public/scoreDay.aspx?cID=412&className=CLUB&dateScoring=201708070)). If the distance would be reduced for the Tps ,the difference would be a multiple of 0.5km as the inner sector is 0.5km (which would add up to 1.5 for 3 Tps). I then analyzed the flight in SeeYou with the respective task and the total task distance was identical with PySoar which again points to a different distance calculation between SeeYou/PySoar and Strepla.
Unfortunately I don't have Strepla myself and just took the task information from the files. Up to now, I haven't found an example with a start-circle on scoringStrepla but in case something like that appears, it would be easy to implement

GliderGeek commented:
it's possible to download a trial version for strepla here:
http://www.strepla.de/StrePla4/english/download/index.htm

i would prefer that all the possible waypoints are implemented, instead of having to update the code frequently

jkretz commented:
Ah, thanks for the hint. I only use SeeYou so I never had a look at their page. I will check that this evening

jkretz commented:
Unfortunately I was not able (or didn't managed it) to get Strepla to write the scoringStrepla task definition into the igc file or somewhere else

GliderGeek commented:
i also gave it a try and didnt succeed. i did however see in strepla that they do have start cylinders, so we should probably make a note somewhere that this is currently no supported. for now, we can leave it as is.

let's get this branch merged, can you look for an appropriate competitionday which we can use for and end to end test? preferably this is a competitionday which a low number of contestants.

will also update the current tests

jkretz commented:
I push two small fixes and the second might also effect the master branch (https issue for soaringspot). For testing I would take this competition [http://www.strepla.de/scs/public/overview.aspx?cID=222](url)

GliderGeek commented:
@jkretz: thanks for the work, the strepla implementation has been merged into master and will be incorporated in the new stand alone build ! 👍
